src/main/java/org/spongepowered/common/resourcepack/SpongeResourcePack.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.resourcepack;
+
+import com.google.common.base.Optional;
+import org.spongepowered.api.resourcepack.ResourcePack;
+
+import java.net.URISyntaxException;
+import java.util.UUID;
+
+public abstract class SpongeResourcePack implements ResourcePack {
+
+    private final Optional<String> hash;
+    private final String id = UUID.randomUUID().toString();
+
+    public static final int HASH_SIZE = 40;
+
+    public SpongeResourcePack(Optional<String> hash) {
+        this.hash = hash;
+    }
+
+    public SpongeResourcePack(String hash) {
+        this(hash.length() == HASH_SIZE ? Optional.of(hash) : Optional.<String>absent());
+    }
+
+    public abstract String getUrlString();
+
+    @Override
+    public String getId() {
+        return this.id;
+    }
+
+    @Override
+    public Optional<String> getHash() {
+        return this.hash;
+    }
+
+    public static SpongeResourcePack create(String url, String hash) throws URISyntaxException {
+        if (url.startsWith("level://")) {
```

Comment:
I don't think it's necessary. I've done `checkNotNull()` on all of the calls in `SpongeResourcePackFactory`.